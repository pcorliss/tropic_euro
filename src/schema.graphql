scalar JSON
scalar JSONObject

type Action {
    key: ID!
}

type Building {
    name: ID!
    description: String
}

type Role {
    name: ID!
    description: String
    doubloons: Int
}

type Plantation {
    type: ID!
    staffed: Boolean
}

type Player {
    name: ID!
    doubloons: Int
}

type Ship {
    spots: Int
    goods: Int
    goodType: String
    size: ID!
}

type GameState {
    id: ID!
    # json: String - This contians private info. do not share
    actionCounter: Int
    roundCounter: Int
    buildings: [Building]
    availableActions(playerId: String): [Action]

    quarries: Int
    colonists: Int
    colonyShip: Int
    victoryPoints: Int

    # lastChange: Int - "Int cannot represent non 32-bit signed integer value: 1656194269745"

    visiblePlantations: [Plantation]

    ships: [Ship]

    availableRoles: [Role]
    currentRole: Role

    players: [Player]
    governor: Player
    currentPlayer: Player
    currentTurnPlayer: Player

    tradingHouse: [String]

    goods: JSONObject
}

type Query {
    hello: String
    gameState(id: String): GameState
}

type Mutation {
   createGame(players: [String]): GameState
   applyAction(gs: String, player: String, key: String): GameState
}