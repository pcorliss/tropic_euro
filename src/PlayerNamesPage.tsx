import React from "react";
import { useQuery, gql } from "@apollo/client";

const PLAYER_NAMES_QUERY = gql`
  query playerNames {
    gameState(id: "aaa") {
      players {
        name
      }
    }
  }
`;

function PlayerNamesPage() {
  const { data, loading, error } = useQuery(PLAYER_NAMES_QUERY);

  if (loading) {
    return <div>loading</div>;
  }

  if (error) {
    return <div>{error.name} - {error.message}</div>;
  }

  return data.gameState.players.map((p: any) => (
    <div key={p.name}>
      <p>
        {p.name}
      </p>
    </div>
  ));
}

export default PlayerNamesPage;