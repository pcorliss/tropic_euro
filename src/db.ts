import DatabaseConstructor, {Database} from "better-sqlite3";

export type Migration = {
    priority: number
    migration: string
}

export class Db {
    static conn: Database;

    static init(options?: any): void {
        options ||= {};
        if (!options["verbose"] && process.env.DB_VERBOSE) {
            options["verbose"] = console.log;
        }
        const dbFile = process.env.DB || ":memory:";
        Db.conn ||= new DatabaseConstructor(dbFile, options);
    }

    static migrate(migrations: Migration[]): void {
        this.init();
        const migrationsCreate = "CREATE TABLE IF NOT EXISTS migrations (priority INTEGER)";
        Db.conn.prepare(migrationsCreate).run();
        const migrationExists = Db.conn.prepare("SELECT 1 FROM migrations WHERE priority = ?");
        const insertMigration = Db.conn.prepare("INSERT INTO migrations VALUES (?)");
        migrations
            .filter((m) => {
                // undefined vs "1" to filter out existing migrations
                return !migrationExists.get(m.priority);
            })
            .sort((a, b) => a.priority - b.priority)
            .forEach((m) => {
                Db.conn.prepare(m.migration).run();
                insertMigration.run(m.priority);
            });
        return;
    }

    static close(): void {
        if (this && this.conn) {
            this.conn.close();
            this.conn = undefined;
        }
        return;
    }
}