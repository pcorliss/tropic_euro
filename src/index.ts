const world = "world";

export function hello(worldStr: string = world): string {
  return `Hello ${worldStr}!`;
}