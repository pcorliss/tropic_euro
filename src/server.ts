import express from "express";
import { graphqlHTTP } from "express-graphql";
import { buildSchema } from "graphql";
import * as fs from "fs";
import { GameState } from "./state/GameState";

export class Server {
  app = express();

  constructor() {
    // Construct a schema, using GraphQL schema language
    const schema = buildSchema(fs.readFileSync("./src/schema.graphql", "utf8"));

    // // The root provides a resolver function for each API endpoint
    const root = {
      hello: () => {
        return "Hello pie!";
      },
      gameState: () => {
        return JSON.stringify(new GameState(["Alice", "Bob", "Carol"]));
      },
    };

    this.app.use("/graphql", graphqlHTTP({
      schema: schema,
      rootValue: root,
      graphiql: true,
    }));
  }
}
