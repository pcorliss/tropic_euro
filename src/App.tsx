import React from 'react';
import { client } from "./ApolloClient/client";
import { ApolloProvider } from '@apollo/client';
import PlayerNamesPage from './PlayerNamesPage';
import './App.css';


function App() {
  return (
    <ApolloProvider client={client}>
      <div>
        <header>Tropic Euro</header>
        <PlayerNamesPage />
      </div>
    </ApolloProvider>
  );
}

export default App;
