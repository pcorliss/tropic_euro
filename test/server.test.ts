import RequestPromise from "request-promise";
import { Server } from "../src/server";
import * as http from "http";

const API = "http://localhost:4999/graphql";

describe("server", () => {
    let s: Server = null;
    let l: http.Server = null;

    beforeAll(async () => {
        s = new Server();
        l = s.app.listen(4999);
    });

    afterAll(async () => {
        l.close();
    });

    it("Hello Pie!", async () => {
        const query = `
            query {
                hello
            }
        `;

        const response = await RequestPromise({method: "POST", uri: API, body: {query}, json: true});
        expect(response.data.hello).toBe("Hello pie!");
        expect(response).toMatchSnapshot();
    });
});