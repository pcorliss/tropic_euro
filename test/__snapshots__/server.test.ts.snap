// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server Hello Pie! 1`] = `
Object {
  "data": Object {
    "hello": "Hello pie!",
  },
}
`;
